### ISSUE TITLE
Can't Opt Out of Linking snprintf()/vsnprintf() in MinGW

### ISSUE THREAD

**Issue by AE1020** [Mbed-TLS/mbedtls#3881]
Building with MinGW [sets a flag MBEDTLS_PLATFORM_HAS_NON_CONFORMING_VSNPRINTF](https://github.com/ARMmbed/mbedtls/blob/9f4a01c33cf72d9ef4d42fc8667afa98805448f9/include/mbedtls/platform.h#L63).

This flag is set in platform.h, after the inclusion of the config.  So there's no way to turn it off.

Once it is on, platform.c [unconditionally defines a helper routine that calls vsnprintf() or vsnprintf_s()](https://github.com/ARMmbed/mbedtls/blob/9f4a01c33cf72d9ef4d42fc8667afa98805448f9/library/platform.c#L136).

An old MinGW configuration I have does not have vsnprintf(), which caused a compiler error to draw attention to the inclusion of the unwanted routine.  Though there seems a [similar issue for snprintf()](https://github.com/ARMmbed/mbedtls/blob/9f4a01c33cf72d9ef4d42fc8667afa98805448f9/library/platform.c#L80)...which it would also be nice to be able to opt out of as well when building under MinGW.

I can just #ifdef these helpers out and avoid the error.  But if something in the config could control it that would be better, to not need to be using a patched mbedTLS.

Here are some settings from my config, where I want to remove all printf()/fprintf()/snprintf()/vsnprintf()'s:

    #define MBEDTLS_PLATFORM_NO_STD_FUNCTIONS

    //#define MBEDTLS_PLATFORM_FPRINTF_ALT
    //#define MBEDTLS_PLATFORM_PRINTF_ALT
    //#define MBEDTLS_PLATFORM_SNPRINTF_ALT
    //#define MBEDTLS_PLATFORM_VSNPRINTF_ALT

    //#define MBEDTLS_PLATFORM_STD_FPRINTF      fprintf /**< Default fprintf to use, can be undefined */
    //#define MBEDTLS_PLATFORM_STD_PRINTF        printf /**< Default printf to use, can be undefined */
    //#define MBEDTLS_PLATFORM_STD_SNPRINTF    snprintf /**< Default snprintf to use, can be undefined */

    //#define MBEDTLS_PLATFORM_FPRINTF_MACRO      fprintf /**< Default fprintf macro to use, can be undefined */
    //#define MBEDTLS_PLATFORM_PRINTF_MACRO        printf /**< Default printf macro to use, can be undefined */
    //#define MBEDTLS_PLATFORM_SNPRINTF_MACRO    snprintf /**< Default snprintf macro to use, can be undefined */
    //#define MBEDTLS_PLATFORM_VSNPRINTF_MACRO    vsnprintf /**< Default vsnprintf macro to use, can be undefined */

So could the definition of the helper function be turned off with these settings?

- Priority: Minor

**OS**  
windows

**mbed TLS build:**  
Version: 9f4a01c33cf72d9ef4d42fc8667afa98805448f9

**Comment by ronald-cron-arm** [comment]
Could adding a check on the MinGW in the following code solve your problem (a check similar to what is done for MSC):
```c
if defined(__MINGW32__) || (defined(_MSC_VER) && _MSC_VER <= 1900)
#define MBEDTLS_PLATFORM_HAS_NON_CONFORMING_SNPRINTF
#define MBEDTLS_PLATFORM_HAS_NON_CONFORMING_VSNPRINTF
#endif
```

**Comment by AE1020** [comment]
> Could adding a check on the MinGW in the following code solve your problem

If you mean checking to see if the MinGW version is old and would not have vsnprintf_s(), it might avoid a compiler error. 

...but my goal *really is* to avoid linking any printf()/snprintf()/vsnprintf() into the binary, at all.  So I would like to avoid these functions that call them unconditionally.

Is there a reason why these couldn't be moved to the config, where they could be turned off?  The default could be the same, but maybe be more descriptive of "you are getting a thing":

    #if defined(__MINGW32__) || (defined(_MSC_VER) && _MSC_VER <= 1900)
    #define MBEDTLS_NEEDS_SNPRINTF_WIN32_SHIM
    #define MBEDTLS_NEEDS_VSNPRINTF_WIN32_SHIM
    #endif

And then in my config, where I "do not want the thing" (along with other disabled things I don't want), I'd just say:

    #if defined(__MINGW32__) || (defined(_MSC_VER) && _MSC_VER <= 1900)
    //#define MBEDTLS_NEEDS_SNPRINTF_WIN32_SHIM
    //#define MBEDTLS_NEEDS_VSNPRINTF_WIN32_SHIM
    #endif

That might be easier than coming up with analysis to look at the flags like MBEDTLS_PLATFORM_STD_SNPRINTF and MBEDTLS_PLATFORM_VSNPRINTF_ALT etc. and try to decide if a combination of them would mean you would or would not want it.

**Comment by gilles-peskine-arm** [comment]
> my goal really is to avoid linking any printf()/snprintf()/vsnprintf() into the binary

The linker should take care of that provided that you don't include any code that calls them. So wouldn't it be enough to add a guard to the definition of `mbedtls_platform_win32_snprintf` with something like
```
#if defined(MBEDTLS_PLATFORM_HAS_NON_CONFORMING_SNPRINTF) && !(__MINGW32__ && __MINGW32_VERSION__ < 42)
```
so that the mbedtls version doesn't get built if the underlying function isn't available, and then you'll get a link error as usual if the function is actually needed?

**Comment by AE1020** [comment]
> The linker should take care of that provided that you don't include any code that calls them. 

...not in my experience:

https://stackoverflow.com/questions/6215782/do-unused-functions-get-optimized-out

**Comment by gilles-peskine-arm** [comment]
In https://github.com/ARMmbed/mbedtls/pull/3453 I tried to rework the code that forces `MBEDTLS_PLATFORM_SNPRINTF_ALT` on, but I ended up with code that fails in some configurations for some reason I don't understand. I do not intend to work on that patch any further, and the structure of Mbed TLS has changed in 3.0 so the patch might need to change significantly.

We would welcome a patch that solves this problem without dropping compatibility with older compiler versions (back to 2013 for 2.2x and 3.0, no clearly defined minimum for mingw but it would at least have to pass our CI).
